### ISSUE TITLE
Precision Error causing models to diverge

### ISSUE THREAD

**Issue by kashyapraval** [keras-team/keras#14852]
A bug
------------------------

### System information

-   **Have I written custom code (as opposed to using a stock example script
    provided in Keras)**: Yes
-   **OS Platform and Distribution (e.g., Linux Ubuntu 16.04)**: Windows 10 but reproducible on google collab
-   **TensorFlow installed from (source or binary)**:
-   **TensorFlow version (use command below)**:v2.5.0-0-ga4dfb8d1a71 2.5.0
-   **Python version**:3.7.10
-   **Bazel version (if compiling from source)**:
-   **GPU model and memory**: google collab GPU, CPU
-   **Exact command to reproduce**: reproduced here - https://colab.research.google.com/drive/1oiWYhSXCLZQHgUOVH-qE9Zr8nBQf6KDE?usp=sharing

You can collect some of this information using our environment capture script:

https://github.com/tensorflow/tensorflow/tree/master/tools/tf_env_collect.sh

You can obtain the TensorFlow version with:

```bash
python -c "import tensorflow as tf; print(tf.version.GIT_VERSION, tf.version.VERSION)"
```

### Describe the problem
2 models( very basic) with same structure and same initial params were trained for just one parameter update with same data.
Using gradient.tape, I verified that the gradients are same and also the learning rate(from callbacks) but the parameter update leads to differences.
### Source code / logs
The following collab notebook highlights just the differences. Could produce the gradient and learning rate consistencies if desired. 
https://colab.research.google.com/drive/1oiWYhSXCLZQHgUOVH-qE9Zr8nBQf6KDE?usp=sharing

**Comment by haifeng-jin** [comment]
Bug verified.
With fixed seed and no randomness in the training process, two cloned models have different gradient updates.
Not sure if the root cause is in the Keras code.

**Comment by kashyapraval** [comment]
with the first gradient update the change is seen only in the weights of the first layer. This further causes all the weights to diverge in subsequent updates

**Comment by haifeng-jin** [comment]
This is a guide for fixing the randomness in Keras.
https://keras.io/getting_started/faq/#how-can-i-obtain-reproducible-results-using-keras-during-development

If you are running on GPU, it is likely that you cannot fix the randomness completely.
